src/statement.ts

```diff
@@ -8,122 +9,344 @@ import { BaseNode, isNode, typeGuard } from "./node";
+  | TryStmt
+  | VariableStmt
+  | WhileStmt;
 
 export function isStmt(a: any): a is Stmt {
   return (
     isNode(a) &&
-    (isBlock(a) ||
+    (isBreakStmt(a) ||
+      isBlockStmt(a) ||
+      isCatchClause(a) ||
+      isContinueStmt(a) ||
       isExprStmt(a) ||
       isForInStmt(a) ||
       isForOfStmt(a) ||
       isIfStmt(a) ||
       isReturn(a) ||
+      isThrowStmt(a) ||
+      isTryStmt(a) ||
       isVariableStmt(a))
   );
 }
 
+export abstract class BaseStmt<
+  Kind extends FunctionlessNode["kind"],
+  Parent extends FunctionlessNode | undefined = BlockStmt | IfStmt
+> extends BaseNode<Kind, Parent> {
+  readonly nodeKind: "Stmt" = "Stmt";
+
+  /**
+   * Node that is prior to this node.
+   */
+  prev: Stmt | undefined;
+  /**
+   * Node that is subsequent to this node.
+   */
+  next: Stmt | undefined;
+}
+
 export const isExprStmt = typeGuard("ExprStmt");
 
-export class ExprStmt extends BaseNode<"ExprStmt"> {
+export class ExprStmt extends BaseStmt<"ExprStmt"> {
   constructor(readonly expr: Expr) {
     super("ExprStmt");
-    expr.parent = this;
+    expr.setParent(this);
+  }
+
+  public clone(): this {
+    return new ExprStmt(this.expr.clone()) as this;
   }
 }
 
 export const isVariableStmt = typeGuard("VariableStmt");
 
+export type VariableStmtParent =
+  | ForInStmt
+  | ForOfStmt
+  | FunctionDecl
+  | FunctionExpr
+  | CatchClause;
+
 export class VariableStmt<
   E extends Expr | undefined = Expr | undefined
-> extends BaseNode<"VariableStmt"> {
+> extends BaseStmt<"VariableStmt", VariableStmtParent> {
   constructor(readonly name: string, readonly expr: E) {
     super("VariableStmt");
     if (expr) {
-      expr.parent = this;
+      expr.setParent(this);
     }
   }
+
+  public clone(): this {
+    return new VariableStmt(this.name, this.expr?.clone()) as this;
+  }
 }
 
-export const isBlock = typeGuard("BlockStmt");
+export const isBlockStmt = typeGuard("BlockStmt");
 
-export class BlockStmt extends BaseNode<"BlockStmt"> {
+export type BlockStmtParent =
+  | CatchClause
+  | DoStmt
+  | ForInStmt
+  | ForOfStmt
+  | FunctionDecl
+  | FunctionExpr
+  | IfStmt
+  | TryStmt
+  | WhileStmt;
+
+export class BlockStmt extends BaseStmt<"BlockStmt", BlockStmtParent> {
   constructor(readonly statements: Stmt[]) {
     super("BlockStmt");
-    let prev = undefined;
-    for (const expr of statements) {
-      expr.parent = this;
-      expr.prev = prev;
-      prev = expr;
+    statements.forEach((stmt, i) => {
+      stmt.setParent(this as never);
+      stmt.prev = i > 0 ? statements[i - 1] : undefined;
+      stmt.next = i + 1 < statements.length ? statements[i + 1] : undefined;
+    });
+  }
+
+  public clone(): this {
+    return new BlockStmt(this.statements.map((stmt) => stmt.clone())) as this;
+  }
+
+  public isFinallyBlock(): this is {
+    parent: TryStmt & {
+      finallyBlock: BlockStmt;
+    };
+  } {
+    return this.parent.kind === "TryStmt" && this.parent.finallyBlock === this;
+  }
```

Comment:
these are really specific helpers? why not just do this as needed or have separate utilities?

Reply:
They are specific about the tree and quite general. I find them to be quite expressive and easy to use - more so than calling a utility. I just `.` onto the object and access the utilities, some of them chain, etc.